File: ios/VideoAnalyzer.swift
import Foundation
import AVFoundation
import Vision
import Photos
import UniformTypeIdentifiers

/// Video analysis modes for NSFW detection
enum VideoScanMode: String {
  /// Quick check - just check beginning, middle, and end (3 frames)
  case quickCheck = "quick_check"

  /// Sampled scan - check at regular intervals (e.g., every 5 seconds)
  case sampled = "sampled"

  /// Thorough scan - use Vision framework to find frames with humans, then ONNX those
  /// Best for newer devices with good Vision performance
  case thorough = "thorough"

  /// Binary search - start at middle, expand outward to find NSFW regions
  /// Good fallback for older devices where Vision is slow
  case binarySearch = "binary_search"

  /// Full scan with short-circuit - check every N seconds until first detection
  case fullWithShortCircuit = "full_short_circuit"
}

/// Result from analyzing a single frame
struct FrameAnalysisResult {
  let timestamp: Double  // seconds
  let isNSFW: Bool
  let confidence: Float
  let detections: [NMS.Detection]
  let processingTime: Int  // ms
}

/// Result from analyzing entire video
struct VideoAnalysisResult {
  let isNSFW: Bool
  let nsfwFrameCount: Int
  let totalFramesAnalyzed: Int
  let firstNSFWTimestamp: Double?  // seconds, nil if no NSFW found
  let nsfwTimestamps: [Double]
  let highestConfidence: Float
  let totalProcessingTime: Int  // ms
  let videoDuration: Double  // seconds
  let scanMode: String
  let humanFramesDetected: Int  // frames where Vision detected humans (for thorough mode)
}

/// Delegate protocol for progress and Live Activity updates
protocol VideoAnalyzerDelegate: AnyObject {
  func videoAnalyzer(_ analyzer: VideoAnalyzer, didUpdateProgress progress: Float)
  func videoAnalyzer(_ analyzer: VideoAnalyzer, didFindNSFWAt timestamp: Double, confidence: Float)
  func videoAnalyzer(_ analyzer: VideoAnalyzer, didComplete result: VideoAnalysisResult)
}

/// Video analyzer using Apple Vision framework + ONNX inference
/// For thorough mode: Uses VNDetectHumanRectanglesRequest to find candidate frames,
/// then runs ONNX NSFW detection only on those frames (much more efficient)
class VideoAnalyzer {

  weak var delegate: VideoAnalyzerDelegate?

  private let detector: ONNXInference
  private let inputSize: Int

  // NSFW class indices from NudeNet model
  private let nsfwClassIndices = Set([2, 3, 4, 6, 14])

  // Configuration
  private var sampleInterval: Double = 5.0
  private let quickCheckPoints: [Double] = [0.0, 0.5, 1.0]  // start, middle, end (as ratio)

  // State for cancellation
  private var isCancelled = false

  init(detector: ONNXInference, inputSize: Int = 640) {
    self.detector = detector
    self.inputSize = inputSize
  }

  func cancel() {
    isCancelled = true
  }

  // MARK: - Public API

  /// Analyze a video from PHAsset
  func analyzeVideo(
    assetId: String,
    mode: VideoScanMode,
    sampleInterval: Double? = nil,
    confidenceThreshold: Float = 0.6
  ) throws -> VideoAnalysisResult {

    isCancelled = false
    self.sampleInterval = sampleInterval ?? 5.0

    guard let phAsset = PHAsset.fetchAssets(withLocalIdentifiers: [assetId], options: nil).firstObject,
          phAsset.mediaType == .video else {
      throw VideoAnalyzerError.assetNotFound
    }

    let avAsset = try getAVAsset(from: phAsset)
    let duration = CMTimeGetSeconds(avAsset.duration)

    NSLog("[VideoAnalyzer] Video: %.1fs, mode: %@, interval: %.1fs", duration, mode.rawValue, self.sampleInterval)

    switch mode {
    case .quickCheck:
      return try analyzeQuickCheck(avAsset: avAsset, duration: duration, confidenceThreshold: confidenceThreshold)

    case .sampled, .fullWithShortCircuit:
      return try analyzeSampled(avAsset: avAsset, duration: duration, mode: mode, confidenceThreshold: confidenceThreshold)

    case .thorough:
      return try analyzeThorough(avAsset: avAsset, duration: duration, confidenceThreshold: confidenceThreshold)

    case .binarySearch:
      return try analyzeBinarySearch(avAsset: avAsset, duration: duration, confidenceThreshold: confidenceThreshold)
    }
  }

  // MARK: - Quick Check (3 frames using AVAssetImageGenerator)

  private func analyzeQuickCheck(
    avAsset: AVAsset,
    duration: Double,
    confidenceThreshold: Float
  ) throws -> VideoAnalysisResult {

    let startTime = Date()
    var results: [FrameAnalysisResult] = []

    let timestamps = quickCheckPoints.map { $0 * max(0.1, duration - 0.1) }

    let generator = createImageGenerator(for: avAsset)

    for (index, timestamp) in timestamps.enumerated() {
      guard !isCancelled else { break }

      delegate?.videoAnalyzer(self, didUpdateProgress: Float(index) / Float(timestamps.count))

      if let result = try? analyzeFrameAtTime(generator: generator, timestamp: timestamp, confidenceThreshold: confidenceThreshold) {
        results.append(result)
        if result.isNSFW {
          delegate?.videoAnalyzer(self, didFindNSFWAt: timestamp, confidence: result.confidence)
        }
      }
    }

    return buildResult(results: results, duration: duration, mode: .quickCheck, startTime: startTime, humanFrames: 0)
  }

  // MARK: - Sampled Analysis (AVAssetImageGenerator at intervals)

  private func analyzeSampled(
    avAsset: AVAsset,
    duration: Double,
    mode: VideoScanMode,
    confidenceThreshold: Float
  ) throws -> VideoAnalysisResult {

    let startTime = Date()
    var results: [FrameAnalysisResult] = []

    // Generate sample timestamps
    var timestamps: [Double] = []
    var t = 0.0
    while t < duration {
      timestamps.append(t)
      t += sampleInterval
    }
    // Include near-end frame
    if timestamps.last ?? 0 < duration - 1.0 {
      timestamps.append(max(0, duration - 0.5))
    }

    NSLog("[VideoAnalyzer] Sampling %d frames at %.1fs intervals", timestamps.count, sampleInterval)

    let generator = createImageGenerator(for: avAsset)

    for (index, timestamp) in timestamps.enumerated() {
      guard !isCancelled else { break }

      delegate?.videoAnalyzer(self, didUpdateProgress: Float(index) / Float(timestamps.count))

      if let result = try? analyzeFrameAtTime(generator: generator, timestamp: timestamp, confidenceThreshold: confidenceThreshold) {
        results.append(result)

        if result.isNSFW {
          delegate?.videoAnalyzer(self, didFindNSFWAt: timestamp, confidence: result.confidence)

          // Short-circuit if requested
          if mode == .fullWithShortCircuit {
            NSLog("[VideoAnalyzer] NSFW at %.1fs, short-circuiting", timestamp)
            break
          }
        }
      }
    }

    return buildResult(results: results, duration: duration, mode: mode, startTime: startTime, humanFrames: 0)
  }

  // MARK: - Thorough Analysis (Vision human detection + ONNX)

  private func analyzeThorough(
    avAsset: AVAsset,
    duration: Double,
    confidenceThreshold: Float
  ) throws -> VideoAnalysisResult {

    let startTime = Date()

    // Step 1: Use Vision to find frames with humans (much faster than ONNX on all frames)
    NSLog("[VideoAnalyzer] Phase 1: Scanning for humans with Vision framework...")

    let humanTimestamps = try findHumanFrames(avAsset: avAsset, duration: duration)

    NSLog("[VideoAnalyzer] Found %d frames with humans", humanTimestamps.count)

    if humanTimestamps.isEmpty {
      // No humans = no NSFW content
      return VideoAnalysisResult(
        isNSFW: false,
        nsfwFrameCount: 0,
        totalFramesAnalyzed: Int(duration / sampleInterval),
        firstNSFWTimestamp: nil,
        nsfwTimestamps: [],
        highestConfidence: 0,
        totalProcessingTime: Int(Date().timeIntervalSince(startTime) * 1000),
        videoDuration: duration,
        scanMode: VideoScanMode.thorough.rawValue,
        humanFramesDetected: 0
      )
    }

    // Step 2: Run ONNX only on frames with humans
    NSLog("[VideoAnalyzer] Phase 2: Running ONNX on %d candidate frames...", humanTimestamps.count)

    var results: [FrameAnalysisResult] = []
    let generator = createImageGenerator(for: avAsset)

    for (index, timestamp) in humanTimestamps.enumerated() {
      guard !isCancelled else { break }

      // Progress: 50% for human detection, 50% for ONNX
      let progress = 0.5 + (Float(index) / Float(humanTimestamps.count)) * 0.5
      delegate?.videoAnalyzer(self, didUpdateProgress: progress)

      if let result = try? analyzeFrameAtTime(generator: generator, timestamp: timestamp, confidenceThreshold: confidenceThreshold) {
        results.append(result)

        if result.isNSFW {
          delegate?.videoAnalyzer(self, didFindNSFWAt: timestamp, confidence: result.confidence)
        }
      }
    }

    return buildResult(results: results, duration: duration, mode: .thorough, startTime: startTime, humanFrames: humanTimestamps.count)
  }

  // MARK: - Binary Search (fallback for older devices)

  private func analyzeBinarySearch(
    avAsset: AVAsset,
    duration: Double,
    confidenceThreshold: Float
  ) throws -> VideoAnalysisResult {

    let startTime = Date()
    var analyzedTimestamps = Set<Double>()
    var results: [FrameAnalysisResult] = []

    let generator = createImageGenerator(for: avAsset)
    let binarySearchWindow: Double = 5.0
    let binarySearchDepth = 3

    // Start with middle ± window
    let middle = duration / 2.0
    var queue: [Double] = []
    for offset in stride(from: -binarySearchWindow, through: binarySearchWindow, by: 1.0) {
      let t = middle + offset
      if t >= 0 && t <= duration {
        queue.append(t)
      }
    }

    var depth = 0
    while depth < binarySearchDepth && !queue.isEmpty {
      guard !isCancelled else { break }

      var nextQueue: [Double] = []

      for timestamp in queue {
        let roundedTime = round(timestamp * 2) / 2
        if analyzedTimestamps.contains(roundedTime) { continue }
        analyzedTimestamps.insert(roundedTime)

        delegate?.videoAnalyzer(self, didUpdateProgress: Float(analyzedTimestamps.count) / Float(min(50, Int(duration / 2))))

        if let result = try? analyzeFrameAtTime(generator: generator, timestamp: timestamp, confidenceThreshold: confidenceThreshold) {
          results.append(result)

          if result.isNSFW {
            delegate?.videoAnalyzer(self, didFindNSFWAt: timestamp, confidence: result.confidence)
            NSLog("[VideoAnalyzer] Binary search: NSFW at %.1fs, expanding", timestamp)

            let before = timestamp - binarySearchWindow
            let after = timestamp + binarySearchWindow
            if before >= 0 { nextQueue.append(before) }
            if after <= duration { nextQueue.append(after) }
          }
        }
      }

      queue = nextQueue
      depth += 1
    }

    return buildResult(results: results, duration: duration, mode: .binarySearch, startTime: startTime, humanFrames: 0)
  }

  /// Use VNDetectHumanRectanglesRequest to find frames containing humans
  private func findHumanFrames(avAsset: AVAsset, duration: Double) throws -> [Double] {
    var humanTimestamps: [Double] = []

    let generator = AVAssetImageGenerator(asset: avAsset)
    generator.appliesPreferredTrackTransform = true
    generator.requestedTimeToleranceBefore = CMTime(seconds: 0.5, preferredTimescale: 600)
    generator.requestedTimeToleranceAfter = CMTime(seconds: 0.5, preferredTimescale: 600)
    // Smaller size for faster Vision processing
    generator.maximumSize = CGSize(width: 320, height: 320)

    var t = 0.0
    var frameIndex = 0
    let totalFrames = Int(duration / sampleInterval) + 1

    while t < duration {
      guard !isCancelled else { break }

      // Progress for phase 1 (0-50%)
      delegate?.videoAnalyzer(self, didUpdateProgress: Float(frameIndex) / Float(totalFrames) * 0.5)

      let requestTime = CMTime(seconds: t, preferredTimescale: 600)
      var actualTime = CMTime.zero

      if let cgImage = try? generator.copyCGImage(at: requestTime, actualTime: &actualTime) {
        // Run Vision human detection
        let hasHuman = detectHumanInFrame(cgImage: cgImage)
        if hasHuman {
          humanTimestamps.append(t)
        }
      }

      t += sampleInterval
      frameIndex += 1
    }

    return humanTimestamps
  }

  /// Quick Vision-based human detection for pre-filtering
  private func detectHumanInFrame(cgImage: CGImage) -> Bool {
    let request = VNDetectHumanRectanglesRequest()
    request.upperBodyOnly = false

    let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])

    do {
      try handler.perform([request])
      if let results = request.results, !results.isEmpty {
        return true
      }
    } catch {
      // Silently continue on Vision errors
    }

    return false
  }

  // MARK: - Helper Methods

  private func getAVAsset(from phAsset: PHAsset) throws -> AVAsset {
    let semaphore = DispatchSemaphore(value: 0)
    var resultAsset: AVAsset?
    var resultError: Error?

    let options = PHVideoRequestOptions()
    options.isNetworkAccessAllowed = false
    options.deliveryMode = .automatic

    PHImageManager.default().requestAVAsset(forVideo: phAsset, options: options) { asset, _, info in
      if let error = info?[PHImageErrorKey] as? Error {
        resultError = error
      } else {
        resultAsset = asset
      }
      semaphore.signal()
    }

    semaphore.wait()

    if let error = resultError { throw error }
    guard let asset = resultAsset else { throw VideoAnalyzerError.failedToLoadVideo }
    return asset
  }

  private func createImageGenerator(for avAsset: AVAsset) -> AVAssetImageGenerator {
    let generator = AVAssetImageGenerator(asset: avAsset)
    generator.appliesPreferredTrackTransform = true
    generator.requestedTimeToleranceBefore = CMTime(seconds: 0.1, preferredTimescale: 600)
    generator.requestedTimeToleranceAfter = CMTime(seconds: 0.1, preferredTimescale: 600)
    generator.maximumSize = CGSize(width: inputSize, height: inputSize)
    return generator
  }

  private func analyzeFrameAtTime(
    generator: AVAssetImageGenerator,
    timestamp: Double,
    confidenceThreshold: Float
  ) throws -> FrameAnalysisResult {

    let frameStart = Date()
    let requestTime = CMTime(seconds: timestamp, preferredTimescale: 600)

    var actualTime = CMTime.zero
    let cgImage = try generator.copyCGImage(at: requestTime, actualTime: &actualTime)

    // Write to temp file for ONNX detector
    let tempURL = FileManager.default.temporaryDirectory
      .appendingPathComponent(UUID().uuidString)
      .appendingPathExtension("jpg")

    defer { try? FileManager.default.removeItem(at: tempURL) }

    // UTType requires iOS 14+ which is our minimum deployment target
    let jpegType = UTType.jpeg.identifier as CFString

    guard let destination = CGImageDestinationCreateWithURL(tempURL as CFURL, jpegType, 1, nil) else {
      throw VideoAnalyzerError.frameExtractionFailed
    }
    CGImageDestinationAddImage(destination, cgImage, nil)
    guard CGImageDestinationFinalize(destination) else {
      throw VideoAnalyzerError.frameExtractionFailed
    }

    let result = try detector.detect(
      imageUri: tempURL.absoluteString,
      confidenceThreshold: confidenceThreshold,
      iouThreshold: 0.45
    )

    let nsfwDetections = result.detections.filter { nsfwClassIndices.contains($0.classIndex) }
    let processingTime = Int(Date().timeIntervalSince(frameStart) * 1000)

    return FrameAnalysisResult(
      timestamp: timestamp,
      isNSFW: !nsfwDetections.isEmpty,
      confidence: nsfwDetections.map { $0.score }.max() ?? 0,
      detections: result.detections,
      processingTime: processingTime
    )
  }

  private func buildResult(
    results: [FrameAnalysisResult],
    duration: Double,
    mode: VideoScanMode,
    startTime: Date,
    humanFrames: Int
  ) -> VideoAnalysisResult {

    let nsfwFrames = results.filter { $0.isNSFW }
    let totalTime = Int(Date().timeIntervalSince(startTime) * 1000)

    let result = VideoAnalysisResult(
      isNSFW: !nsfwFrames.isEmpty,
      nsfwFrameCount: nsfwFrames.count,
      totalFramesAnalyzed: results.count,
      firstNSFWTimestamp: nsfwFrames.first?.timestamp,
      nsfwTimestamps: nsfwFrames.map { $0.timestamp }.sorted(),
      highestConfidence: results.map { $0.confidence }.max() ?? 0,
      totalProcessingTime: totalTime,
      videoDuration: duration,
      scanMode: mode.rawValue,
      humanFramesDetected: humanFrames
    )

    delegate?.videoAnalyzer(self, didComplete: result)
    delegate?.videoAnalyzer(self, didUpdateProgress: 1.0)

    return result
  }
}

// MARK: - Errors

enum VideoAnalyzerError: Error, LocalizedError {
  case assetNotFound
  case failedToLoadVideo
  case frameExtractionFailed
  case detectorNotReady
  case cancelled

  var errorDescription: String? {
    switch self {
    case .assetNotFound: return "Video asset not found"
    case .failedToLoadVideo: return "Failed to load video from Photos library"
    case .frameExtractionFailed: return "Failed to extract frame from video"
    case .detectorNotReady: return "ONNX detector not initialized"
    case .cancelled: return "Video analysis was cancelled"
    }
  }
}

File: android/src/main/java/com/visionml/VideoAnalyzer.kt
package com.visionml

import android.content.ContentResolver
import android.content.Context
import android.graphics.Bitmap
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.provider.MediaStore
import android.util.Log
import kotlin.math.max
import kotlin.math.round

/**
 * Video analyzer for NSFW content detection
 * Ported from iOS VideoAnalyzer.swift
 *
 * Uses MediaMetadataRetriever for frame extraction + ONNXInference for detection
 */
class VideoAnalyzer(
    private val context: Context,
    private val detector: ONNXInference,
    private val inputSize: Int = 640
) {
    companion object {
        private const val TAG = "VideoAnalyzer"
        private val NSFW_CLASS_INDICES = setOf(2, 3, 4, 6, 14)
    }

    enum class ScanMode(val value: String) {
        /** Quick check - just check beginning, middle, and end (3 frames) */
        QUICK_CHECK("quick_check"),
        /** Sampled scan - check at regular intervals (e.g., every 5 seconds) */
        SAMPLED("sampled"),
        /** Thorough scan - use ML Kit to find frames with humans, then ONNX those */
        THOROUGH("thorough"),
        /** Binary search - start at middle, expand outward to find NSFW regions */
        BINARY_SEARCH("binary_search"),
        /** Full scan with short-circuit - check every N seconds until first detection */
        FULL_SHORT_CIRCUIT("full_short_circuit");

        companion object {
            fun fromString(value: String): ScanMode {
                return values().find { it.value == value } ?: SAMPLED
            }
        }
    }

    /** Result from analyzing a single frame */
    data class FrameAnalysisResult(
        val timestamp: Double,  // seconds
        val isNSFW: Boolean,
        val confidence: Float,
        val detections: List<NMS.Detection>,
        val processingTime: Int  // ms
    )

    /** Result from analyzing entire video */
    data class VideoAnalysisResult(
        val isNSFW: Boolean,
        val nsfwFrameCount: Int,
        val totalFramesAnalyzed: Int,
        val firstNSFWTimestamp: Double?,
        val nsfwTimestamps: List<Double>,
        val highestConfidence: Float,
        val totalProcessingTime: Int,  // ms
        val videoDuration: Double,  // seconds
        val scanMode: String,
        val humanFramesDetected: Int
    )

    /** Delegate for progress updates */
    interface Delegate {
        fun onProgressUpdate(progress: Float)
        fun onNSFWFound(timestamp: Double, confidence: Float)
        fun onComplete(result: VideoAnalysisResult)
    }

    var delegate: Delegate? = null
    private var isCancelled = false

    // Configuration
    private var sampleInterval = 5.0
    private val quickCheckPoints = listOf(0.0, 0.5, 1.0)  // start, middle, end (as ratio)

    fun cancel() {
        isCancelled = true
    }

    /**
     * Analyze a video from content URI or media ID
     */
    fun analyzeVideo(
        assetId: String,
        mode: ScanMode,
        sampleInterval: Double = 5.0,
        confidenceThreshold: Float = 0.6f
    ): VideoAnalysisResult {
        isCancelled = false
        this.sampleInterval = sampleInterval

        val retriever = MediaMetadataRetriever()

        try {
            // Try to open the video
            val uri = resolveAssetId(assetId)
            retriever.setDataSource(context, uri)

            val durationMs = retriever.extractMetadata(MediaMetadataRetriever.METADATA_KEY_DURATION)
                ?.toLongOrNull() ?: 0L
            val duration = durationMs / 1000.0

            Log.d(TAG, "Video: ${duration}s, mode: ${mode.value}, interval: ${this.sampleInterval}s")

            return when (mode) {
                ScanMode.QUICK_CHECK -> analyzeQuickCheck(retriever, duration, confidenceThreshold)
                ScanMode.SAMPLED, ScanMode.FULL_SHORT_CIRCUIT -> analyzeSampled(retriever, duration, mode, confidenceThreshold)
                ScanMode.THOROUGH -> analyzeThorough(retriever, duration, confidenceThreshold)
                ScanMode.BINARY_SEARCH -> analyzeBinarySearch(retriever, duration, confidenceThreshold)
            }
        } finally {
            retriever.release()
        }
    }

    /**
     * Resolve asset ID to content URI
     * Supports:
     * - content:// URIs directly
     * - MediaStore IDs (numeric)
     * - file:// URIs
     */
    private fun resolveAssetId(assetId: String): Uri {
        return when {
            assetId.startsWith("content://") -> Uri.parse(assetId)
            assetId.startsWith("file://") -> Uri.parse(assetId)
            assetId.all { it.isDigit() } -> {
                // Numeric ID - assume MediaStore video ID
                Uri.withAppendedPath(MediaStore.Video.Media.EXTERNAL_CONTENT_URI, assetId)
            }
            else -> Uri.parse(assetId)
        }
    }

    // MARK: - Quick Check (3 frames)

    private fun analyzeQuickCheck(
        retriever: MediaMetadataRetriever,
        duration: Double,
        confidenceThreshold: Float
    ): VideoAnalysisResult {
        val startTime = System.currentTimeMillis()
        val results = mutableListOf<FrameAnalysisResult>()

        val timestamps = quickCheckPoints.map { it * max(0.1, duration - 0.1) }

        for ((index, timestamp) in timestamps.withIndex()) {
            if (isCancelled) break

            delegate?.onProgressUpdate(index.toFloat() / timestamps.size)

            val result = analyzeFrameAtTime(retriever, timestamp, confidenceThreshold)
            if (result != null) {
                results.add(result)
                if (result.isNSFW) {
                    delegate?.onNSFWFound(timestamp, result.confidence)
                }
            }
        }

        return buildResult(results, duration, ScanMode.QUICK_CHECK, startTime, 0)
    }

    // MARK: - Sampled Analysis

    private fun analyzeSampled(
        retriever: MediaMetadataRetriever,
        duration: Double,
        mode: ScanMode,
        confidenceThreshold: Float
    ): VideoAnalysisResult {
        val startTime = System.currentTimeMillis()
        val results = mutableListOf<FrameAnalysisResult>()

        // Generate sample timestamps
        val timestamps = mutableListOf<Double>()
        var t = 0.0
        while (t < duration) {
            timestamps.add(t)
            t += sampleInterval
        }
        // Include near-end frame
        if ((timestamps.lastOrNull() ?: 0.0) < duration - 1.0) {
            timestamps.add(max(0.0, duration - 0.5))
        }

        Log.d(TAG, "Sampling ${timestamps.size} frames at ${sampleInterval}s intervals")

        for ((index, timestamp) in timestamps.withIndex()) {
            if (isCancelled) break

            delegate?.onProgressUpdate(index.toFloat() / timestamps.size)

            val result = analyzeFrameAtTime(retriever, timestamp, confidenceThreshold)
            if (result != null) {
                results.add(result)

                if (result.isNSFW) {
                    delegate?.onNSFWFound(timestamp, result.confidence)

                    // Short-circuit if requested
                    if (mode == ScanMode.FULL_SHORT_CIRCUIT) {
                        Log.d(TAG, "NSFW at ${timestamp}s, short-circuiting")
                        break
                    }
                }
            }
        }

        return buildResult(results, duration, mode, startTime, 0)
    }

    // MARK: - Thorough Analysis (ML Kit human detection + ONNX)

    private fun analyzeThorough(
        retriever: MediaMetadataRetriever,
        duration: Double,
        confidenceThreshold: Float
    ): VideoAnalysisResult {
        val startTime = System.currentTimeMillis()

        // Step 1: Use ML Kit to find frames with humans (if available)
        // For now, fall back to sampled mode since ML Kit requires additional setup
        Log.d(TAG, "Phase 1: Scanning for humans (falling back to sampled)...")

        // TODO: Integrate MLKitAnalyzer for human detection
        // For now, analyze all sampled frames
        val timestamps = mutableListOf<Double>()
        var t = 0.0
        while (t < duration) {
            timestamps.add(t)
            t += sampleInterval
        }

        Log.d(TAG, "Analyzing ${timestamps.size} frames...")

        val results = mutableListOf<FrameAnalysisResult>()

        for ((index, timestamp) in timestamps.withIndex()) {
            if (isCancelled) break

            val progress = index.toFloat() / timestamps.size
            delegate?.onProgressUpdate(progress)

            val result = analyzeFrameAtTime(retriever, timestamp, confidenceThreshold)
            if (result != null) {
                results.add(result)
                if (result.isNSFW) {
                    delegate?.onNSFWFound(timestamp, result.confidence)
                }
            }
        }

        return buildResult(results, duration, ScanMode.THOROUGH, startTime, timestamps.size)
    }

    // MARK: - Binary Search

    private fun analyzeBinarySearch(
        retriever: MediaMetadataRetriever,
        duration: Double,
        confidenceThreshold: Float
    ): VideoAnalysisResult {
        val startTime = System.currentTimeMillis()
        val analyzedTimestamps = mutableSetOf<Double>()
        val results = mutableListOf<FrameAnalysisResult>()

        val binarySearchWindow = 5.0
        val binarySearchDepth = 3

        // Start with middle ± window
        val middle = duration / 2.0
        var queue = mutableListOf<Double>()
        var offset = -binarySearchWindow
        while (offset <= binarySearchWindow) {
            val t = middle + offset
            if (t >= 0 && t <= duration) {
                queue.add(t)
            }
            offset += 1.0
        }

        var depth = 0
        while (depth < binarySearchDepth && queue.isNotEmpty()) {
            if (isCancelled) break

            val nextQueue = mutableListOf<Double>()

            for (timestamp in queue) {
                val roundedTime = round(timestamp * 2) / 2
                if (roundedTime in analyzedTimestamps) continue
                analyzedTimestamps.add(roundedTime)

                delegate?.onProgressUpdate(analyzedTimestamps.size.toFloat() / 50f.coerceAtMost((duration / 2).toFloat()))

                val result = analyzeFrameAtTime(retriever, timestamp, confidenceThreshold)
                if (result != null) {
                    results.add(result)

                    if (result.isNSFW) {
                        delegate?.onNSFWFound(timestamp, result.confidence)
                        Log.d(TAG, "Binary search: NSFW at ${timestamp}s, expanding")

                        val before = timestamp - binarySearchWindow
                        val after = timestamp + binarySearchWindow
                        if (before >= 0) nextQueue.add(before)
                        if (after <= duration) nextQueue.add(after)
                    }
                }
            }

            queue = nextQueue
            depth++
        }

        return buildResult(results, duration, ScanMode.BINARY_SEARCH, startTime, 0)
    }

    // MARK: - Helper Methods

    private fun analyzeFrameAtTime(
        retriever: MediaMetadataRetriever,
        timestamp: Double,
        confidenceThreshold: Float
    ): FrameAnalysisResult? {
        val frameStart = System.currentTimeMillis()
        val timestampUs = (timestamp * 1_000_000).toLong()  // Convert to microseconds

        val bitmap = retriever.getFrameAtTime(timestampUs, MediaMetadataRetriever.OPTION_CLOSEST_SYNC)
            ?: return null

        return try {
            val result = detector.detectBitmap(bitmap, confidenceThreshold, 0.45f)

            val nsfwDetections = result.detections.filter { it.classIndex in NSFW_CLASS_INDICES }
            val processingTime = (System.currentTimeMillis() - frameStart).toInt()

            FrameAnalysisResult(
                timestamp = timestamp,
                isNSFW = nsfwDetections.isNotEmpty(),
                confidence = nsfwDetections.maxOfOrNull { it.score } ?: 0f,
                detections = result.detections,
                processingTime = processingTime
            )
        } finally {
            bitmap.recycle()
        }
    }

    private fun buildResult(
        results: List<FrameAnalysisResult>,
        duration: Double,
        mode: ScanMode,
        startTime: Long,
        humanFrames: Int
    ): VideoAnalysisResult {
        val nsfwFrames = results.filter { it.isNSFW }
        val totalTime = (System.currentTimeMillis() - startTime).toInt()

        val result = VideoAnalysisResult(
            isNSFW = nsfwFrames.isNotEmpty(),
            nsfwFrameCount = nsfwFrames.size,
            totalFramesAnalyzed = results.size,
            firstNSFWTimestamp = nsfwFrames.firstOrNull()?.timestamp,
            nsfwTimestamps = nsfwFrames.map { it.timestamp }.sorted(),
            highestConfidence = results.maxOfOrNull { it.confidence } ?: 0f,
            totalProcessingTime = totalTime,
            videoDuration = duration,
            scanMode = mode.value,
            humanFramesDetected = humanFrames
        )

        delegate?.onComplete(result)
        delegate?.onProgressUpdate(1f)

        return result
    }
}
